lib/elixir/lib/registry.ex

```diff
@@ -653,6 +653,70 @@ defmodule Registry do
   end
 
   @doc """
+  Unregister entries for a given key matching a pattern.
+
+  ## Examples
+
+  For unique registries it's the same as unregister:
+
+
+  For duplicate registries:
+
+      iex> Registry.start_link(:duplicate, Registry.DuplicateUnregisterMatchTest)
+      iex> Registry.register(Registry.DuplicateUnregisterMatchTest, "hello", :world_a)
+      iex> Registry.register(Registry.DuplicateUnregisterMatchTest, "hello", :world_b)
+      iex> Registry.register(Registry.DuplicateUnregisterMatchTest, "hello", :world_c)
+      iex> Registry.keys(Registry.DuplicateUnregisterMatchTest, self())
+      ["hello", "hello", "hello"]
+      iex> Registry.unregister_match(Registry.DuplicateUnregisterMatchTest, "hello", :world_a)
+      :ok
+      iex> Registry.keys(Registry.DuplicateUnregisterMatchTest, self())
+      ["hello", "hello"]
+      iex> Registry.lookup(Registry.DuplicateUnregisterMatchTest, "hello")
+      [:world_b, :world_c]
+  """
+  def unregister_match(registry, key, pattern) do
+    self = self()
+
+    {kind, partitions, key_ets, pid_ets, listeners} = info!(registry)
+    {key_partition, pid_partition} = partitions(kind, key, self, partitions)
+    key_ets = key_ets || key_ets!(registry, key_partition)
+    {pid_server, pid_ets} = pid_ets || pid_ets!(registry, pid_partition)
+
+    # Remove first from the key_ets because in case of crashes
+    # the pid_ets will still be able to clean up. The last step is
+    # to clean if we have no more entries.
+    # spec = {key, {self, pattern}}
+    select_spec = [{{key, {self, pattern}},[],[true]}]
+    total   = :ets.select_count(key_ets, spec) |> IO.inspect
+    case :ets.select_delete(key_ets, {key, {self, pattern}}) do
+      # We deleted everything, we can just delete the object
+      ^total ->
+        true = :ets.delete_object(pid_ets, {self, key, key_ets})
+
+      deleted ->
+        remaining = total - deleted
+        fake_key_ets = -1 * remaining
+
+        :ets.insert(pid_ets, {self, key, fake_key_ets})
```

Comment:
We need to include the `key_ets` here so would be `{key_ets, remaining}` or similar because if the caller gets an exit signal between `delete_object` and `insert` then there is no `key_ets` for the GC process to cleanup the key. Otherwise the GC process needs to be able to hash to get the `key_ets`.